docs/proposals/local-registries.md

```diff
@@ -0,0 +1,165 @@
+
+When a `BuildRun` is instantiated, the image specs that declare `local: true` are mutated to inject
+the local image registry host.
+
+### User Stories [optional]
+
+As an application developer
+I would like to declare that images in builds reference a local image registry
+So that I do not have to provide host information in the pull spec
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Obtaining local registry information
+
+First, clusters which deploy a local image registry must provide the releavant host info by
+publishing it in the `local-registry-hosting`
+[ConfigMap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry#the-local-registry-hosting-configmap).
+The build operator/controller must have RBAC which allows it to read the `ConfigMap` and obtain the
+registry host information. For `Build` and `BuildRun` objects, the `HostFromClusterNetwork` value
+should be used as the hostname.
+
+If the `local-registry-hosting` ConfigMap is created, updated, or deleted, the build controller
+and/or operator needs to react accordingly.
+
+#### Declaring an image as local
+
+`Image` references in the API will be updated to have the `local` field. An example using the 
+`output` spec:
+
+```yaml
+spec:
+  output:
+    image: mynamespace/myapp:latest
+    local: true
+```
+
+When an image is declared as `local`, the build controller prepends the local image registry
+hostname to the provided image reference.
+
+### Risks and Mitigations
+
+App developers use `local: true` on clusters that don't have a local registry defined, end up
+pulling images from `docker.io` or an image registry in container runtime's unqualified search
+path.
+
+Local image registry configurations can be altered while the build controller is running.
+
+## Design Details
+
+### Test Plan
+
+1. Deploy the build operator/controller on a cluster that populates the `local-registry-hosting`
+   ConfigMap OR as cluster-admin, manually populate the `local-registry-hosting` ConfigMap.
+2. Create a `Build` and/or `BuildRun` which references a local image.
+3. Ensure that a build can push an image referencing the local image spec.
+
+### Graduation Criteria
+
+**Dev Preview** - we can simply define an environment variable to set the local registry hostname.
+
+**Tech Preview/GA** - fully read the `local-registry-hosting` ConfigMap, with the ability to update
+the build controller if this ConfigMap changes.
+
+#### Examples
+
+These are generalized examples to consider, in addition to the aforementioned [maturity
+levels][maturity-levels].
+
+### Upgrade / Downgrade Strategy
+
+Upgrades add a new field to the API. On downgrade the `local` attribute should be ignored.
+
+### Version Skew Strategy
+
+N/A
+
+## Implementation History
+
+2020-07-24: Initial proposal
+
+## Drawbacks
+
+TBD
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
```

Comment:
Pasting some of my open questions from a previous comment in here, while here feels like the best place:

I see the [configmap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry#the-local-registry-hosting-configmap) approach as a best practice. However that does not mean that this best practice will be applicable for every cloud providers. Some questions:
- I took a look at the [initial code](https://github.com/redhat-developer/build/compare/master...adambkaplan:local-registries-impl?expand=1), it seems to me that we can already push an image into a local registry. This is how is done already for e2e tests. Therefore the `Support Local Image Registries` feature is already available.
- Based on the above comment, is this proposal more related to the fact around using the `configmap` approach? If this is true, we might need a more generic configmap approach(_e.g. the docs enforce the following `The namespace must be kube-public`, probably we do not want this in all cases_) 
- Do we need this configmap approach? What I envision by using this approach is the following:
  1. Users would be able to push an image to a local registry by setting the `spec.output.local` to true
  2. At the moment without the configmap  approach, what users need to do is to set the proper local registry endpoint in the `spec.output.image`
- Between 1 and 2 from above, whats the aggregated value?